feat: Config builder

WillLillis opened the pull request:
I am greatly excited by the config improvements coming from #158. These changes do make the config a bit more complex though, which may be confusing/hard to track in large projects/ for new users. This PR adds a small CLI program that prompts a user through creating a `.asm-lsp.toml` config file, which should help manage that complexity. ~~I'll add some more details to this PR description (and some example gifs) once it's closer to being merged, but I wanted to open this up early for the sake of visibility.~~

~~Currently, the config builder program lives as a separate binary crate within the project. I think I'd like to change this to run as a subcommand of the main program (assuming it doesn't bloat the binary size by too much). Some documentation for this also needs to be added to the project's README.~~

Here's an example gif. I don't currently have a large assembly project to demonstrate this on, so I'm just going to run the tool within the `asm-lsp` repo selecting some nonsense options:

![ConfigBuilder](https://github.com/user-attachments/assets/59c8c9ee-9d03-45c0-b8ce-c14fe934115f)

(The flickering only shows up in the recording, I'm not entirely sure what's causing that)

The builder provides validation for selected options in the following ways:

- Any enum field specified (i.e. arch or assembler) will be valid
- project paths will point to valid directories or files on the system. After selection, the builder translates the absolute path to one relative to the project's root directory (for portability reasons)
- For any compiler specified, a check is made to ensure it is a valid path to an executable file, or that it can be found on `$PATH`.
- Project configs are checked for redundant paths

Closes #161

WillLillis commented on asm-lsp-cfg/src/main.rs:
```diff
@@ -0,0 +1,434 @@
+    fn try_from(value: GenerateArgs) -> Result<Self, std::string::String> {
+        let output_path = {
+            if value.global_cfg {
+                config_dir().ok_or_else(|| String::from("Failed to detect config directory, try specifying it manually with `--output_dir`"))?
```
Note to self, need to push the correct sub path here

WillLillis commented on asm-lsp-cfg/src/main.rs:
```diff
@@ -0,0 +1,434 @@
+            }
+            !(!exists || !is_file)
+        } else {
+            let find_cmd = if cfg!(target_os = "windows") {
```
I had originally wrote code to traverse over all the directories in $PATH here to search for the inputted compiler name. It took several seconds of an invalid name was given, so I opted for this approach instead. If it can be optimized instead, I'd happily replace this hack.

Relevant code snippet:

```rust
let mut found = false;
'path_var_loop: for entry in path_var.split(':').filter(|e| !e.is_empty()) {
    let dir_path = PathBuf::from(entry);
    if let Ok(dir_reader) = std::fs::read_dir(&dir_path) {
        if dir_reader
            .filter_map(std::result::Result::ok)
            .filter(|entry| entry.path().is_file())
            .map(|file| file.file_name())
            .find(|name| name.eq(comp.as_str()))
            .is_some_and(|_| true)
        {
            found = true;
            break 'path_var_loop;
        }
    }
}

if !found {
    println!("Warning: Unable to find provided compiler on $PATH");
}
```

Ultra-Code replied:
```rust
pub fn is_executable(cmd: &str) -> bool {
    use std::env;
    use std::fs;
    use std::path::Path;
    // Check if a path points to an executable file
    let is_executable = {
        |path: &Path| -> bool {
            if path.is_file() {
                #[cfg(unix)]
                {
                    // On Unix, check the `x` bit
                    use std::os::unix::fs::PermissionsExt;
                    let metadata = fs::metadata(path).unwrap();
                    metadata.permissions().mode() & 0o111 != 0
                }
                #[cfg(windows)]
                {
                    // On Windows, check for common executable extensions
                    let extensions = ["exe", "cmd", "bat", "com"];
                    if let Some(ext) = path.extension().and_then(|s| s.to_str()) {
                        return extensions.contains(&ext);
                    }
                    false
                }
            } else {
                false
            }
        }
    };

    // Get the PATH environment variable
    let path_var = env::var_os("PATH").unwrap();

    for path in env::split_paths(&path_var) {
        let full_path = path.join(cmd);
        if is_executable(&full_path) {
            return true;
        }
    }
    false
}
```

This might help, try it out

WillLillis replied:
This is a huge improvement, thank you!

WillLillis commented on asm-lsp-cfg/src/main.rs:
```diff
@@ -0,0 +1,434 @@
+        let mut flags = Vec::new();
+        loop {
+            let flag: String = Input::with_theme(&ColorfulTheme::default())
+                .with_prompt("Compile Flag: (Enter an empty string to stop)")
```
Probably need to reword this a bit

WillLillis commented:
@lu-zero I'm curious how well this fits in with your use case for dav1d (very cool project by the way!). If you have the time to test this out, I'd really like to address any issues/pain points you run into. :)

WillLillis commented:
Rebased and added some checks to ensure multiple projects couldn't have the same path. I'm planning on merging this in tomorrow night unless I spot some other issues.

Ultra-Code commented:
Looks good, Impressive work 🔥 🚀 .
